File: proof_builder.c
/* SPDX-FileCopyrightText: 2025 Rhett Creighton
 * SPDX-License-Identifier: Apache-2.0
 */

#define _POSIX_C_SOURCE 200809L
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/wait.h>

#define MAX_PATHS 1000
#define MAX_PATH_NAME 256
#define MAX_CMD_LEN 8192

typedef struct {
    char name[MAX_PATH_NAME];
    int exists;
} path_config_t;

typedef struct {
    char source_file[256];
    char output_binary[256];
    path_config_t paths[MAX_PATHS];
    int path_count;
    int proving_build;
    int tracing_build;
} build_config_t;

/* Extract all SECURE_CODE_POINT names from source */
int extract_code_points(const char* filename, build_config_t* config) {
    FILE* fp = fopen(filename, "r");
    if (!fp) {
        perror("Failed to open source file");
        return -1;
    }
    
    char line[1024];
    config->path_count = 0;
    
    while (fgets(line, sizeof(line), fp)) {
        char* pos = strstr(line, "SECURE_CODE_POINT(");
        if (pos) {
            pos += strlen("SECURE_CODE_POINT(");
            char* end = strchr(pos, ',');
            if (end) {
                int len = end - pos;
                if (len < MAX_PATH_NAME - 1) {
                    strncpy(config->paths[config->path_count].name, pos, len);
                    config->paths[config->path_count].name[len] = '\0';
                    config->paths[config->path_count].exists = 1; /* Default enabled */
                    config->path_count++;
                }
            }
        }
    }
    
    fclose(fp);
    return config->path_count;
}

/* Build command with path definitions */
void build_proof_binary(build_config_t* config) {
    char cmd[MAX_CMD_LEN];
    int offset = 0;
    
    /* Start with compiler and flags */
    offset += snprintf(cmd + offset, MAX_CMD_LEN - offset, 
                      "gcc -std=c99 -I./include");
    
    /* Add build mode */
    if (config->proving_build) {
        offset += snprintf(cmd + offset, MAX_CMD_LEN - offset, " -DPROVING_BUILD");
    } else if (config->tracing_build) {
        offset += snprintf(cmd + offset, MAX_CMD_LEN - offset, " -DTRACING_BUILD");
    }
    
    /* Add all path definitions */
    for (int i = 0; i < config->path_count; i++) {
        offset += snprintf(cmd + offset, MAX_CMD_LEN - offset, 
                          " -DPATH_%s_EXISTS=%d",
                          config->paths[i].name, 
                          config->paths[i].exists);
    }
    
    /* Add source and output */
    offset += snprintf(cmd + offset, MAX_CMD_LEN - offset, 
                      " %s -o %s -lm",
                      config->source_file,
                      config->output_binary);
    
    printf("Building: %s\n", cmd);
    system(cmd);
}

/* Enable specific paths by name */
void enable_paths(build_config_t* config, const char* path_list) {
    /* First disable all */
    for (int i = 0; i < config->path_count; i++) {
        config->paths[i].exists = 0;
    }
    
    /* Parse comma-separated list */
    char* list_copy = strdup(path_list);
    char* token = strtok(list_copy, ",");
    
    while (token) {
        /* Find and enable this path */
        for (int i = 0; i < config->path_count; i++) {
            if (strcmp(config->paths[i].name, token) == 0) {
                config->paths[i].exists = 1;
                break;
            }
        }
        token = strtok(NULL, ",");
    }
    
    free(list_copy);
}

/* Disable specific paths by name */
void disable_paths(build_config_t* config, const char* path_list) {
    char* list_copy = strdup(path_list);
    char* token = strtok(list_copy, ",");
    
    while (token) {
        /* Find and disable this path */
        for (int i = 0; i < config->path_count; i++) {
            if (strcmp(config->paths[i].name, token) == 0) {
                config->paths[i].exists = 0;
                break;
            }
        }
        token = strtok(NULL, ",");
    }
    
    free(list_copy);
}

int main(int argc, char* argv[]) {
    if (argc < 3) {
        fprintf(stderr, "Usage: %s <source.c> <output> [options]\n", argv[0]);
        fprintf(stderr, "Options:\n");
        fprintf(stderr, "  --proving        Build for proof mode (exits at points)\n");
        fprintf(stderr, "  --tracing        Build for trace mode (records paths)\n");
        fprintf(stderr, "  --enable <paths> Enable only these paths (comma-separated)\n");
        fprintf(stderr, "  --disable <paths> Disable these paths (comma-separated)\n");
        return 1;
    }
    
    build_config_t config = {0};
    strcpy(config.source_file, argv[1]);
    strcpy(config.output_binary, argv[2]);
    
    /* Extract all code points from source */
    if (extract_code_points(config.source_file, &config) < 0) {
        return 1;
    }
    
    printf("Found %d secure code points\n", config.path_count);
    
    /* Parse options */
    for (int i = 3; i < argc; i++) {
        if (strcmp(argv[i], "--proving") == 0) {
            config.proving_build = 1;
        } else if (strcmp(argv[i], "--tracing") == 0) {
            config.tracing_build = 1;
        } else if (strcmp(argv[i], "--enable") == 0 && i + 1 < argc) {
            enable_paths(&config, argv[++i]);
        } else if (strcmp(argv[i], "--disable") == 0 && i + 1 < argc) {
            disable_paths(&config, argv[++i]);
        }
    }
    
    /* Build the binary */
    build_proof_binary(&config);
    
    return 0;
}

File: proof_builder.py
#!/usr/bin/env python3
"""Build unique binaries to prove code path reachability."""

import hashlib
import json
import os
import subprocess
import sys
from pathlib import Path

class SecureProofBuilder:
    def __init__(self, source_file, proof_manifest):
        self.source_file = source_file
        self.proof_manifest = proof_manifest
        self.build_dir = Path("proof_builds")
        self.build_dir.mkdir(exist_ok=True)
        
    def generate_sha3_tag(self, name):
        """Generate a unique 256-bit SHA3 tag for a code point."""
        sha3 = hashlib.sha3_256()
        sha3.update(name.encode('utf-8'))
        return sha3.hexdigest()
        
    def extract_all_paths(self):
        """Extract all SECURE_CODE_POINT names from source file."""
        paths = set()
        with open(self.source_file, 'r') as f:
            content = f.read()
            # Find all SECURE_CODE_POINT(NAME, ...) occurrences
            import re
            matches = re.findall(r'SECURE_CODE_POINT\((\w+),', content)
            paths.update(matches)
        return sorted(list(paths))
        
    def build_proof_binary(self, test_name, enabled_paths, disabled_paths):
        """Build a binary with specific paths enabled/disabled."""
        # Get all paths from source file
        all_paths = self.extract_all_paths()
        
        defines = [f"-DPROVING_BUILD"]
        
        # Define all paths (default to 0 if not mentioned)
        for path in all_paths:
            if path in enabled_paths:
                defines.append(f"-DPATH_{path}_EXISTS=1")
            elif path in disabled_paths:
                defines.append(f"-DPATH_{path}_EXISTS=0")
            else:
                # Default to disabled if not mentioned
                defines.append(f"-DPATH_{path}_EXISTS=0")
            
        # Generate SHA3 tags
        tags = {}
        for path in enabled_paths:
            tags[path] = self.generate_sha3_tag(path)
            
        # Build command
        output = self.build_dir / f"proof_{test_name}"
        cmd = [
            "gcc",
            "-std=c99",
            "-I./include",
            *defines,
            self.source_file,
            "-o", str(output),
            "-lm"
        ]
        
        print(f"Building proof binary: {test_name}")
        print(f"  Enabled paths: {enabled_paths}")
        print(f"  Disabled paths: {disabled_paths}")
        print(f"  Command: {' '.join(cmd)}")
        
        result = subprocess.run(cmd, capture_output=True, text=True)
        
        if result.returncode != 0:
            print(f"Build failed:\n{result.stderr}")
            return None
            
        return {
            "binary": str(output),
            "enabled_paths": enabled_paths,
            "disabled_paths": disabled_paths,
            "tags": tags
        }
        
    def run_proof_binary(self, binary_info, inputs):
        """Run a proof binary with specific inputs."""
        binary = binary_info["binary"]
        
        print(f"\nRunning proof: {binary}")
        print(f"  Inputs: {inputs}")
        
        # Run with deterministic inputs
        env = os.environ.copy()
        for key, value in inputs.items():
            env[key] = str(value)
            
        result = subprocess.run([binary], env=env, capture_output=True, text=True)
        
        if "SECURE_EXIT:" in result.stdout:
            # Extract exit point
            for line in result.stdout.split('\n'):
                if line.startswith("SECURE_EXIT:"):
                    exit_point = line.split(":")[1].strip()
                    print(f"  Reached secure exit: {exit_point}")
                    return exit_point
                    
        print(f"  No secure exit reached")
        return None
        
    def generate_proof_index(self, results):
        """Generate index of what each secure exit implies."""
        index = {
            "proofs": [],
            "unreachable_paths": [],
            "reachable_paths": {}
        }
        
        for test_name, result in results.items():
            proof = {
                "test": test_name,
                "binary": result["binary"]["binary"],
                "inputs": result["inputs"],
                "exit_point": result["exit_point"],
                "enabled_paths": result["binary"]["enabled_paths"],
                "disabled_paths": result["binary"]["disabled_paths"],
                "implies": result.get("implies", "")
            }
            
            index["proofs"].append(proof)
            
            # Track reachability
            if result["exit_point"]:
                point = result["exit_point"]
                if point not in index["reachable_paths"]:
                    index["reachable_paths"][point] = []
                index["reachable_paths"][point].append(test_name)
                
        # Find universally unreachable paths
        all_paths = set()
        for proof in index["proofs"]:
            all_paths.update(proof["enabled_paths"])
            all_paths.update(proof["disabled_paths"])
            
        reached_paths = set(index["reachable_paths"].keys())
        index["unreachable_paths"] = list(all_paths - reached_paths)
        
        return index
        
    def run_proof_suite(self):
        """Run complete proof suite from manifest."""
        with open(self.proof_manifest) as f:
            manifest = json.load(f)
            
        results = {}
        
        for test in manifest["tests"]:
            test_name = test["name"]
            
            # Build binary
            binary_info = self.build_proof_binary(
                test_name,
                test.get("enabled_paths", []),
                test.get("disabled_paths", [])
            )
            
            if not binary_info:
                continue
                
            # Run with various inputs
            exit_point = None
            for input_set in test["inputs"]:
                exit_point = self.run_proof_binary(binary_info, input_set)
                if exit_point:
                    break
                    
            results[test_name] = {
                "binary": binary_info,
                "inputs": test["inputs"],
                "exit_point": exit_point,
                "implies": test.get("implies", "")
            }
            
        # Generate index
        index = self.generate_proof_index(results)
        
        # Save index
        with open(self.build_dir / "proof_index.json", "w") as f:
            json.dump(index, f, indent=2)
            
        print("\n=== PROOF SUMMARY ===")
        print(f"Total tests: {len(results)}")
        print(f"Reachable paths: {len(index['reachable_paths'])}")
        print(f"Unreachable paths: {len(index['unreachable_paths'])}")
        
        if index['unreachable_paths']:
            print("\nPROVEN UNREACHABLE:")
            for path in index['unreachable_paths']:
                print(f"  - {path}")
                
        return index


if __name__ == "__main__":
    if len(sys.argv) != 3:
        print(f"Usage: {sys.argv[0]} <source_file> <proof_manifest.json>")
        sys.exit(1)
        
    builder = SecureProofBuilder(sys.argv[1], sys.argv[2])
    builder.run_proof_suite()